### PR TITLE
Adding bunch of unary foreach APIs

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47385 Added foreach_trunc, foreahc_reciprocal, foreach_sigmoid APIs
* #47384 Added foreach_frac API
* **#47875 Adding bunch of unary foreach APIs**

Implementing several unary operators for _foreach_ APIs.
### Planned list of ops 
- [x]  abs
- [x]  acos
- [x]  asin
- [x]  atan
- [x]  ceil
- [x]  cos
- [x]  cosh
- [x]  erf
- [x]  erfc
- [x]  exp
- [x]  expm1
- [x]  floor
- [x]  log
- [x]  log10
- [x]  log1p
- [x]  log2
- [ ]  frac
- [x]  neg
- [ ]  reciprocal
- [x]  round
- [ ]  rsqrt
- [ ]  sigmoid
- [x]  sin
- [x]  sinh
- [x]  sqrt
- [x]  tan
- [x]  tanh
- [ ]  trunc
- [x]  lgamma
- [ ]  digamma
- [ ]  erfinv
- [ ]  sign
- [ ]  mvlgamma
- [ ]  clamp
- [ ]  clamp_min
- [ ]  clamp_max

### Perf results
```
----------------- OP:  sin  -----------------
  Median: 998.79 us
  300.84 us

----------------- OP:  abs  -----------------
  Median: 1.19 ms
  294.97 us

----------------- OP:  acos  -----------------
  Median: 982.30 us
  299.40 us

----------------- OP:  asin  -----------------
  Median: 1.16 ms
  298.09 us

----------------- OP:  atan  -----------------
  Median: 986.92 us
  295.64 us

----------------- OP:  ceil  -----------------
  Median: 1.17 ms
  297.25 us

----------------- OP:  cos  -----------------
  Median: 972.72 us
  294.41 us

----------------- OP:  cosh  -----------------
  Median: 1.17 ms
  294.97 us

----------------- OP:  erf  -----------------
  Median: 1.17 ms
  297.02 us

----------------- OP:  erfc  -----------------
  Median: 1.14 ms
  299.23 us

----------------- OP:  exp  -----------------
  Median: 1.15 ms
  298.79 us

----------------- OP:  expm1  -----------------
  Median: 1.17 ms
  291.79 us

----------------- OP:  floor  -----------------
  Median: 1.17 ms
  293.51 us

----------------- OP:  log  -----------------
  Median: 1.13 ms
  318.01 us

----------------- OP:  log10  -----------------
  Median: 987.17 us
  295.57 us

----------------- OP:  log1p  -----------------
  Median: 1.13 ms
  297.15 us

----------------- OP:  log2  -----------------
  Median: 974.21 us
  295.01 us

----------------- OP:  frac  -----------------
  Median: 1.15 ms
  296.01 us

----------------- OP:  neg  -----------------
  Median: 1.13 ms
  294.98 us

----------------- OP:  reciprocal  -----------------
  Median: 1.16 ms
  293.69 us

----------------- OP:  round  -----------------
  Median: 1.12 ms
  297.48 us

----------------- OP:  sigmoid  -----------------
  Median: 1.13 ms
  296.53 us

----------------- OP:  sin  -----------------
  Median: 991.02 us
  295.78 us

----------------- OP:  sinh  -----------------
  Median: 1.15 ms
  295.70 us

----------------- OP:  sqrt  -----------------
  Median: 1.17 ms
  297.75 us

----------------- OP:  tan  -----------------
  978.20 us
  297.99 us

----------------- OP:  tanh  -----------------
  Median: 967.84 us
  297.29 us

----------------- OP:  trunc  -----------------
  Median: 1.14 ms
  298.72 us

----------------- OP:  lgamma  -----------------
  Median: 1.14 ms
  317.53 us
```

### Script

```

import torch
import torch.optim as optim
import torch.nn as nn
import torchvision
import torch.utils.benchmark as benchmark_utils

inputs = [torch.rand(3, 200, 200, device="cuda") for _ in range(100)]

def main():
    for op in [
            "sin", "abs", "acos", "asin", "atan", "ceil",
            "cos", "cosh", "erf", "erfc",
            "exp", "expm1", "floor", "log",
            "log10", "log1p", "log2", "frac",
            "neg", "reciprocal", "round",
            "sigmoid", "sin", "sinh", "sqrt",
            "tan", "tanh", "trunc", "lgamma"
        ]:
        print("\n\n----------------- OP: ", op, " -----------------")
        stmt = "[torch.{op}(t) for t in inputs]"
        timer = benchmark_utils.Timer(
            stmt=stmt.format(op = op),
            globals=globals(),
            label="str(optimizer)",
        )
        print(f"autorange:\n{timer.blocked_autorange()}\n\n")


        stmt = "torch._foreach_{op}(inputs)"
        timer_mta = benchmark_utils.Timer(
            stmt=stmt.format(op = op),
            globals=globals(),
            label="str(optimizer_mta)",
        )
        print(f"autorange:\n{timer_mta.blocked_autorange()}\n\n")

if __name__ == "__main__":
    main()

```

Differential Revision: [D24948801](https://our.internmc.facebook.com/intern/diff/D24948801)

### CODE REVIEW

File: test/test_foreach.py [hunk 1]
```diff
@@ -149,13 +163,106 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
...
+
+                        with self.assertRaisesRegex(RuntimeError, r"In-place abs is not supported for complex tensors."):
+                            torch._foreach_abs_(tensors1)
+                        break
+                    else:
+                        self.assertEqual(res, expected)
+                else:
+                    if dtype == torch.complex64:
+                        expected = [torch.abs(tensors1[i]).to(torch.complex64) for i in range(N)]
+                        self.assertEqual(res, expected)
```
Comment (ashishfarmer) [inline test/test_foreach.py:252]: res computed as `res = torch._foreach_abs(tensors1)` has real datatype for the tensors in the list, whereas the above line makes each tensor in the `expected` list a complex tensor. 
That test is failing on ROCm because of this type difference. Anything I am missing or is behaving differently than it should on ROCm?

Comment (izdeby) [reply test/test_foreach.py:252]: in case of complex input, the result will be of real type. Currently, foreach APIs dont support different return type. As a fix we are pushing complex to a slow path.

File: test/test_foreach.py [hunk 2]
```diff
@@ -149,13 +163,106 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
...
+
+                        with self.assertRaisesRegex(RuntimeError, r"In-place abs is not supported for complex tensors."):
+                            torch._foreach_abs_(tensors1)
+                        break
+                    else:
+                        self.assertEqual(res, expected)
+                else:
+                    if dtype == torch.complex64:
+                        expected = [torch.abs(tensors1[i]).to(torch.complex64) for i in range(N)]
```
Comment (ngimel) [inline test/test_foreach.py:251]: wait, how does this work now? `res` is real, but you are converting `expected` to complex?

Comment (izdeby) [reply test/test_foreach.py:251]: @ngimel , yes, because foreach APIs dont allow returning different type than input one. This will be fixed once i roll out (this week) the stack where we move this kind of cases to a slow path.

Comment (izdeby) [reply test/test_foreach.py]: It was wrong. I wasn't pushing abs with complex input to a slow path

**Review by ngimel — APPROVED** [review]
